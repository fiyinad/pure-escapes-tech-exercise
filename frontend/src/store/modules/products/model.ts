export interface Product {
    id: number;
    name: string;
    customerPrice: number;
    cost: number;
}

export interface IProductsDomain {
    value: Array<Product>;
    loading: boolean;
    error: boolean;
}

export const initialState: IProductsDomain = {
    value: [],
    loading: false,
    error: false
};
