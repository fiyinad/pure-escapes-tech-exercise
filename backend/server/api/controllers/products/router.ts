import express from 'express';
import controller from './controller';
export default express
  .Router()
  .post('/', controller.create)
  .post('/summary', controller.summary)
  .get('/', controller.all)
  .get('/:id', controller.byId);
