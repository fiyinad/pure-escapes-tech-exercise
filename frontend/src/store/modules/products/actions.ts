import { Product } from "./model";

export const GET_ALL_REQUEST = 'GET_ALL_REQUEST';
export const GET_ALL_SUCCESS = 'GET_ALL_SUCCESS';
export const GET_ALL_FAILURE = 'GET_ALL_FAILURE';
export const GET_ALL = 'GET_ALL';
export const GET_SINGLE = 'GET_SINGLE';

export type GetAllRequestAction = ReturnType<typeof getAllRequestAction>;
export const getAllRequestAction = () => ({
  type: GET_ALL_REQUEST as typeof GET_ALL_REQUEST
});

export type GetAllSuccessAction = ReturnType<typeof getAllSuccessAction>;
export const getAllSuccessAction = () => ({
  type: GET_ALL_SUCCESS as typeof GET_ALL_SUCCESS
});

export type GetAllFailureAction = ReturnType<typeof getAllFailureAction>;
export const getAllFailureAction = () => ({
  type: GET_ALL_FAILURE as typeof GET_ALL_FAILURE
});

export type GetAllAction = ReturnType<typeof getAllAction>;
export const getAllAction = (payload: Array<Product>) => ({
  type: GET_ALL as typeof GET_ALL,
  payload,
});

export type GetSingleAction = ReturnType<typeof getSingleAction>;
export const getSingleAction = (payload: Product) => ({
  type: GET_SINGLE as typeof GET_SINGLE,
  payload 
});

export type ProductsAction =
  | GetAllRequestAction
  | GetAllSuccessAction
  | GetAllFailureAction
  | GetAllAction
  | GetSingleAction;
