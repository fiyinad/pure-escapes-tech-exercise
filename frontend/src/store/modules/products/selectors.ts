import { IProductsDomain } from './model';
import { createSelector } from 'reselect';

export const productsDomain = (state: any): IProductsDomain => state.products;

export const productsSelector = createSelector(
    productsDomain,
    (domain: IProductsDomain): IProductsDomain['value'] => domain.value
);

export const productsLoadingSelector = createSelector(
    productsDomain,
    (domain: IProductsDomain): IProductsDomain['loading'] => domain.loading
);
