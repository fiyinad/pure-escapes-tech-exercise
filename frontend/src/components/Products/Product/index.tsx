import React from 'react';
import { Button, ListItem, ListItemSecondaryAction, ListItemText, Typography } from '@material-ui/core';
import { Product as ProductModel } from '../../../store/modules/products/model';

export interface IProductProps {
    product: ProductModel;
    onAdd: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>, item: ProductModel) => void;
    onRemove: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>, item: ProductModel) => void;
  }

const Product: React.FC<IProductProps> = props => {
    return (
        <>
            <ListItem>
                <ListItemText>
                    <Typography variant="body1" color='textPrimary'>
                        {props.product?.name}
                    </Typography>
                </ListItemText>
                <ListItemSecondaryAction>
                    <Button 
                        variant="outlined" 
                        aria-label="add" 
                        size="small" 
                        color="primary" 
                        onClick={(e) => props.onAdd(e, props.product)}>
                        Add
                    </Button>
                </ListItemSecondaryAction>
            </ListItem>
        </>
    );
};

export default Product;