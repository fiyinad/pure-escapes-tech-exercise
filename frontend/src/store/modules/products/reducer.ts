import { IProductsDomain, initialState, Product } from './model';
import * as Actions from './actions';

const upsert = (arr: Array<Product>, object: Product) => {
    const i = arr.findIndex(_object => _object.id === object.id);
    if (i > -1) arr[i] = object; 
    else arr.push(object);
    return arr;
}

const productsReducer = (state: IProductsDomain = initialState, action: Actions.ProductsAction) => {
    switch (action.type) {
        case Actions.GET_ALL:
            return { ...state, value: [...action.payload] };

        case Actions.GET_SINGLE:

            return { ...state, value: upsert([...state.value], action.payload) };

        case Actions.GET_ALL_REQUEST:
            return { ...state, loading: true, error: false };

        case Actions.GET_ALL_SUCCESS:
            return { ...state, loading: false, error: false };

        case Actions.GET_ALL_FAILURE:
            return { ...state, loading: false, error: true };

        default:
            return state;
    }
};

export default productsReducer;
