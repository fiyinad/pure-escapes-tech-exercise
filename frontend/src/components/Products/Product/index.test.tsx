import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Product, { IProductProps } from './index';
import { Product as ProductModel } from '../../../store/modules/products/model';

describe('Product', () => {
    it('Renders without errors in default state', () => {
        // arrange, 
        const props = {} as IProductProps;
        
        // act
        const component = shallow(<Product {...props} />);

        // assert
        expect(toJson(component)).toMatchSnapshot();
    });

    it('Renders without errors with populated state', () => {
        // arrange, 
        const product: ProductModel = { "id": 1, "name": "Soup", "customerPrice": 199, "cost": 186 };
        const props = {
            product,
            onAdd: jest.fn(),
            onRemove: jest.fn(),
        } as IProductProps;
        
        // act
        const component = shallow(<Product {...props} />);

        // assert
        expect(toJson(component)).toMatchSnapshot();
    });
});