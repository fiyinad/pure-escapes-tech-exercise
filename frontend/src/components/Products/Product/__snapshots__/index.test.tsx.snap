// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Renders without errors in default state 1`] = `
<Fragment>
  <WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItemText))>
      <WithStyles(ForwardRef(Typography))
        color="textPrimary"
        variant="body1"
      />
    </WithStyles(ForwardRef(ListItemText))>
    <WithStyles(ForwardRef(ListItemSecondaryAction))>
      <WithStyles(ForwardRef(Button))
        aria-label="add"
        color="primary"
        onClick={[Function]}
        size="small"
        variant="outlined"
      >
        Add
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(ListItemSecondaryAction))>
  </WithStyles(ForwardRef(ListItem))>
</Fragment>
`;

exports[`Product Renders without errors with populated state 1`] = `
<Fragment>
  <WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItemText))>
      <WithStyles(ForwardRef(Typography))
        color="textPrimary"
        variant="body1"
      >
        Soup
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(ListItemText))>
    <WithStyles(ForwardRef(ListItemSecondaryAction))>
      <WithStyles(ForwardRef(Button))
        aria-label="add"
        color="primary"
        onClick={[Function]}
        size="small"
        variant="outlined"
      >
        Add
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(ListItemSecondaryAction))>
  </WithStyles(ForwardRef(ListItem))>
</Fragment>
`;
